Lib/socket.py

```diff
@@ -728,6 +728,76 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
     else:
         raise error("getaddrinfo returns an empty list")
 
+def has_dual_stack():
+    """Return True if the kernel supports creating a socket which
+    can handle both AF_INET and AF_INET6 (IPv4 / IPv6) connections.
+    """
+    if not has_ipv6 \
+            or not hasattr(_socket, 'AF_INET6') \
+            or not hasattr(_socket, 'IPV6_V6ONLY'):
+        return False
+    try:
+        with socket(AF_INET6, SOCK_STREAM) as sock:
+            if not sock.getsockopt(IPPROTO_IPV6, IPV6_V6ONLY):
+                return True
+            else:
+                sock.setsockopt(IPPROTO_IPV6, IPV6_V6ONLY, False)
+                return True
+    except error:
+        return False
+
+def create_server_sock(address, *, dual_stack=has_dual_stack(), queue_size=5,
+                       reuse_addr=os.name == 'posix' and sys.platform != 'cygwin'):
```

Comment:
In `asyncio` the flag is called `reuse_address`. Maybe use the same name here for sake of consistency?

`queue_size` is named `backlog` with the same meaning.
I suspect `5` as default is too small. In `asyncio` the parameter default value is `100`. For blocking sockets high value is even more important than for async case.

Also `asyncio` `loop.create_server()` supports `reuse_port()` flag.

Reply:
By the way, it would be useful to find out if `asyncio`'s variant can be rewritten to benefit from this new function.

Reply:
asyncio implementation is here: https://github.com/python/cpython/blob/master/Lib/asyncio/base_events.py#L1300-L1415

Reply:
> In asyncio the flag is called reuse_address. Maybe use the same name here for sake of consistency?

Mmm. I think that for an asyncio app this function is fundamentally useless because it will use `asyncio` API. You're not supposed to import socket where you import asyncio, so I'm not sure how much consistency matters in this case. 

> queue_size is named backlog with the same meaning.

I will rename it to backlog

> I suspect 5 as default is too small. In asyncio the parameter default value is 100. For blocking sockets high value is even more important than for async case.

Good point. I took a look at some projects. Tornado uses 128, Twisted uses 50, curio uses 100, Trio (and Golang) uses a big size (but never exceding 65536):
https://github.com/python-trio/trio/blob/cfa82d8abc42a0bd56be9bc788686cb22590ace6/trio/_highlevel_open_tcp_listeners.py#L10
One may argue whether we may want to set a default kwarg for `socket.listen()` and also `socketserver` module (but those are separate issues).
CC-ing @njsmith who I suspect may be interested in this. 

> Also asyncio loop.create_server() supports reuse_port() flag.

Interesting. I never needed `SO_REUSEPORT` myself but it appears it's useful so it probably deserves its own argument: 
https://medium.com/uckey/the-behaviour-of-so-reuseport-addr-1-2-f8a440a35af6